File: 백준/Platinum/1725. 히스토그램/히스토그램.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		int[] height = new int[n + 1];

		for (int i = 0; i < n; i++) {
			height[i] = Integer.parseInt(br.readLine());
		}

		ArrayDeque<Integer> stack = new ArrayDeque<>();

		int result = 0;

		for (int i = 0; i <= n; i++) {
			while (!stack.isEmpty() && height[stack.peek()] >= height[i]) {
				int h = height[stack.pop()];
				int w;

				if (stack.isEmpty()) {
					w = i;
				} else {
					w = i - stack.peek() - 1;
				}
				result = Math.max(result, h * w);
			}
			stack.push(i);
		}

		System.out.println(result);
	}
}

File: 백준/Platinum/1725. 히스토그램/히스토그램.py
import sys

n = int(sys.stdin.readline().rstrip())
height = [int(sys.stdin.readline().rstrip()) for _ in range(n)] + [0]
stack = []
result = 0

for i in range(n + 1):
    while stack and height[stack[-1]] >= height[i]:
        h = height[stack.pop()]
        w = 0

        if not stack:
            w = i
        else:
            w = i - stack[-1] - 1

        result = max(result, h * w)

    stack.append(i)

print(result)
